### ISSUE TITLE
(Lambda):  Add SignedCode.fromAsset option

### ISSUE THREAD

**Issue by dontirun** [aws/aws-cdk#13286]
Related to https://github.com/aws/aws-cdk/pull/12656 , https://github.com/aws/aws-cdk/issues/12216

Created a Lambda `SignedCode.fromAsset`  option that takes local code uploads it to S3 and signs it using a specified AWS signer profile

### Use Case

With PR https://github.com/aws/aws-cdk/pull/12656 Lambda now supports a code signing configuration. However if the signing config is set to `Enforce` and local code (inline or from asset path) is provided the deployment will fail as the code has not been signed

```typescript
      const signingProfile = new signer.SigningProfile(stack, 'SigningProfile', {
        platform: signer.Platform.AWS_LAMBDA_SHA384_ECDSA,
      });

      const codeSigningConfig = new lambda.CodeSigningConfig(stack, 'CodeSigningConfig', {
        signingProfiles: [signingProfile],
        untrustedArtifactOnDeployment: lambda.UntrustedArtifactOnDeployment.ENFORCE,
      });

      new lambda.Function(stack, 'MyLambda', {
        code: new lambda.Code.fromAsset(...),
        handler: 'index.handler',
        runtime: lambda.Runtime.NODEJS_10_X,
        codeSigningConfig,
      });
  ```

this feature would enable usage of local code and signing of the the code given permissions to the signing profile

### Proposed Solution

Having an option like
```
     new lambda.Function(stack, 'MyLambda', {
        code: new lambda.SignedCode.fromAsset(...),
        handler: 'index.handler',
        runtime: lambda.Runtime.NODEJS_10_X,
        codeSigningConfig,
      });
```
would solve this issue

### Other
1. This is [how the SAM CLI does it](https://github.com/aws/aws-sam-cli/blob/develop/samcli/lib/package/code_signer.py)
2. The bootstrap bucket would need to have [versioning enabled](https://github.com/aws/aws-cli/blob/5aa599949f60b6af554fd5714d7161aa272716f7/awscli/examples/signer/start-signing-job.rst)

* [ ] :wave: I may be able to implement this feature request
* [X] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

**Comment by hedrall** [comment]
I read the code around `aws-lmabda.code`, and I thought the following part is a good way to extend it.

https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/core/lib/asset-staging.ts#L161-L169

'aws-lambda-nodejs` builds the code with esbuild, but passes the following values to the constructor of aws-lambda.

```typescript
code: Code.fromAsset(path.dirname(options.depsLockFilePath), {
  assetHashType: cdk.AssetHashType.OUTPUT,
  Bundling: new Bundling(options),
})
```

Building process is executed locally by default, but if esbuild command not available, executed in compatible docker container.

As for signatures, is it possible to rely on the local environment?
 Also, which tool should I rely on, AWS CLI, AWS SDK, etc.?

**Comment by jogold** [comment]
Looks like something for [`cdk-assets`](https://github.com/aws/aws-cdk/tree/master/packages/cdk-assets)?

**Comment by dontirun** [comment]
> 
> 
> I read the code around `aws-lmabda.code`, and I thought the following part is a good way to extend it.
> 
> https://github.com/aws/aws-cdk/blob/1fcdb6daf931147b8f33facb8ab9c9f80e5c9eee/packages/%40aws-cdk/core/lib/asset-staging.ts#L161-L169
> 
> In the case of `aws-lambda-nodejs`, the following values can be passed to the constructor of `aws-lambda.Function` to build the code with esbuild.
> 
> ```ts
> code: Code.fromAsset(path.dirname(options.depsLockFilePath), {
>   assetHashType: cdk.AssetHashType.OUTPUT,
>   Bundling: new Bundling(options),
> })
> ```
> 
> Building process is executed locally by default, but if esbuild command not available in the environment, executed in compatible docker container.
> 
> As for signatures, is it possible to rely on the local environment?
> Also, which tool should I rely on, AWS CLI, AWS SDK, etc.?
> 
> (Once we have a policy for implementation, I would like to start implementing it.)

I don't think we need to touch the local bundling. The signing need to occur after the object has been uploaded to the S3 bucket (I believe the upload is done [here](https://github.com/aws/aws-cdk/blob/05a998065b3333854715c456b20b7cc5d5daac67/packages/cdk-assets/lib/private/handlers/files.ts#L51). ) I believe an optimal approach is to follow what the [SAM CLI does](https://github.com/aws/aws-sam-cli/blob/develop/samcli/lib/package/code_signer.py) with their python code
1. Add an optional parameter with the profile name(without the version)
1. [Start the signing job](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Signer.html#startSigningJob-property)
2. Use the [built in waiter](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Signer.html#waitFor-property) for the signing job to complete.

In addition, the [CDK bootstrap bucket](https://github.com/aws/aws-cdk/blob/b96efd8aaa143845b9fe315a9ee1e8398c4d83c2/packages/aws-cdk/lib/api/bootstrap/bootstrap-template.yaml#L140) would need to be changed to have versioning enabled ( since Signer requires a versioned object in S3 to start a signing job)

**Comment by hedrall** [comment]
Thank you. I misunderstood it as something to be signed locally.

I understand that we should start signing job after s3.upload.
[https://github.com/aws/aws-cdk/blob/05a998065b3333854715c456b20b7cc5d5daac67/packages/cdk-assets/lib/private/handlers/files.ts#L51-L56](https://github.com/aws/aws-cdk/blob/05a998065b3333854715c456b20b7cc5d5daac67/packages/cdk-assets/lib/private/handlers/files.ts#L51-L56)

To specify asset that need to be singed, should we path profile name of signer here?,,

[https://github.com/aws/aws-cdk/blob/05a998065b3333854715c456b20b7cc5d5daac67/packages/%40aws-cdk/aws-s3-assets/lib/asset.ts#L147-L151](https://github.com/aws/aws-cdk/blob/05a998065b3333854715c456b20b7cc5d5daac67/packages/%40aws-cdk/aws-s3-assets/lib/asset.ts#L147-L151)

**Comment by hedrall** [comment]
Is this a good place to add parameters related to signature settings?

https://github.com/aws/aws-cdk/blob/9910f77b3644e350254f797a0a992c490ea41289/packages/%40aws-cdk/cloud-assembly-schema/lib/cloud-assembly/metadata-schema.ts#L26-L49
